core/src/main/java/com/linecorp/armeria/client/circuitbreaker/KeyedCircuitBreakerMapping.java

```diff
@@ -20,66 +20,59 @@
+ * breaker factory.
  */
 final class KeyedCircuitBreakerMapping implements CircuitBreakerMapping {
 
-    static final CircuitBreakerMapping hostMapping =
-            new KeyedCircuitBreakerMapping(MappingKey.HOST, (host, method) -> CircuitBreaker.of(host));
+    static final CircuitBreakerMapping hostMapping = new KeyedCircuitBreakerMapping(
+            new CircuitBreakerMapping.Builder().perHost(),
+            (host, method, path) -> CircuitBreaker.of(host));
 
     private final ConcurrentMap<String, CircuitBreaker> mapping = new ConcurrentHashMap<>();
 
-    private final MappingKey mappingKey;
-    private final BiFunction<String, String, ? extends CircuitBreaker> factory;
+    private final boolean isPerHost;
+    private final boolean isPerMethod;
+    private final boolean isPerPath;
+    private final CircuitBreakerFactory factory;
 
     /**
-     * Creates a new {@link KeyedCircuitBreakerMapping} with the given {@link MappingKey} and
+     * Creates a new {@link KeyedCircuitBreakerMapping} with the given {@link CircuitBreakerMapping.Builder} and
      * {@link CircuitBreaker} factory.
      */
-    KeyedCircuitBreakerMapping(MappingKey mappingKey,
-                               BiFunction<String, String, ? extends CircuitBreaker> factory) {
-        this.mappingKey = requireNonNull(mappingKey, "mappingKey");
+    KeyedCircuitBreakerMapping(CircuitBreakerMapping.Builder mappingBuilder, CircuitBreakerFactory factory) {
+        requireNonNull(mappingBuilder, "mappingBuilder");
+        isPerHost = mappingBuilder.isPerHost();
+        isPerMethod = mappingBuilder.isPerMethod();
+        isPerPath = mappingBuilder.isPerPath();
         this.factory = requireNonNull(factory, "factory");
     }
 
     @Override
     public CircuitBreaker get(ClientRequestContext ctx, Request req) throws Exception {
-        final String key;
-        final String host;
-        final String method;
-        switch (mappingKey) {
-            case HOST:
-                key = host = host(ctx);
-                method = null;
-                break;
-            case METHOD:
-                host = null;
-                key = method = method(ctx);
-                break;
-            case HOST_AND_METHOD:
-                host = host(ctx);
-                method = method(ctx);
-                key = host + '#' + method;
-                break;
-            default:
-                // should never reach here.
-                throw new Error();
+        if (!isPerHost && !isPerMethod && !isPerPath) {
+            // should never reach here.
+            throw new Error();
         }
+        final String host = isPerHost ? host(ctx) : null;
+        final String method = isPerMethod ? method(ctx) : null;
+        final String path = isPerPath ? path(ctx) : null;
+        final String key =
+                (isPerHost ? host : "") + '#' + (isPerMethod ? method : "") + '#' + (isPerPath ? path : "");
```

Comment:
I think we can drop `#` if one of `isPer*` flags is false.
Guava `Joiner` could be useful for this case.
```java
static Joiner keyJoiner = Joiner.on('#').skipNulls();
String key = joiner.join(host, method, path);
```

Reply:
Done through java streams.